### PR TITLE
Fix issue 21243 - Allow lambdas to return auto ref

### PR DESCRIPTION
**Pull request by pbackus**
Spec PR: https://github.com/dlang/dlang.org/pull/3410

### CODE REVIEW

**Review by ljmf00 — APPROVED** [review]
LGTM

**Comment by SixthDot** [comment]
coverage is not 100%, please add a test to fix that.

**Comment by pbackus** [comment]
Auto-tester failure is a segfault in [this static destructor][1] that occurs non-deterministically when running `std.concurrency`'s test suite. Probably a race between the main thread writing to the `__gshared` variable `scheduler` and test threads reading from it in their static destructors. Almost certainly unrelated to this PR.

[1]: https://github.com/dlang/phobos/blob/v2.100.2/std/concurrency.d#L266-L269

**Comment by ljmf00** [comment]
Is this, https://github.com/dlang/phobos/pull/8579, the fix you to that race conditions?

**Comment by pbackus** [comment]
@ljmf00 Yes.

**Comment by pbackus** [comment]
Force-pushed to re-run now that the race condition has (hopefully) been fixed.

**Review by thewilsonator — APPROVED** [review]

**Comment by pbackus** [comment]
Buildkite is having availability issues; force-pushed to re-run again.

File: compiler/src/dmd/parse.d [hunk 1]
```diff
@@ -4857,6 +4857,10 @@ class Parser(AST, Lexer = dmd.lexer.Lexer) : Lexer
                     token.value == TOK.identifier && peekNext() == TOK.goesTo ||
                     token.value == TOK.ref_ && peekNext() == TOK.leftParenthesis &&
                         skipAttributes(peekPastParen(peek(&token)), &tk) &&
+                        (tk.value == TOK.goesTo || tk.value == TOK.leftCurly) ||
+                    token.value == TOK.auto_ && peekNext() == TOK.ref_ &&
```
Comment (Geod24) [inline compiler/src/dmd/parse.d:4861]: Ident issue

File: compiler/src/dmd/parse.d [hunk 2]
```diff
@@ -4995,7 +5001,20 @@ class Parser(AST, Lexer = dmd.lexer.Lexer) : Lexer
...
+                nextToken();
+                if (token.value == TOK.ref_)
+                {
+                    // function auto ref (parameters) { statements... }
+                    // delegate auto ref (parameters) { statements... }
+                    stc = STC.auto_ | STC.ref_;
+                    nextToken();
+                }
+                else
+                    error("`auto` can only be used as part of `auto ref` for function literal return values");
```
Comment (Geod24) [inline compiler/src/dmd/parse.d:5015]: Should we error here ? `function auto` is just like `function`. So either don't error or add an `errorSupplemental` telling the user to remove `auto` / `auto` is already the default behavior ?

File: compiler/src/dmd/parse.d [hunk 3]
```diff
@@ -5023,6 +5042,20 @@ class Parser(AST, Lexer = dmd.lexer.Lexer) : Lexer
...
+                nextToken();
+                if (token.value == TOK.ref_)
+                {
+                    // auto ref (parameters) => expression
+                    // auto ref (parameters) { statements... }
+                    stc = STC.auto_ | STC.ref_;
+                    nextToken();
+                }
+                else
+                    error("`auto` can only be used as part of `auto ref` for function literal return values");
```
Comment (Geod24) [inline compiler/src/dmd/parse.d:5056]: Ditto

**Review by Geod24 — COMMENTED** [review]
Nice

File: compiler/src/dmd/parse.d (hunk 1 above)
Comment (pbackus) [reply compiler/src/dmd/parse.d:4861]: The non-standard indentation here is intentional, and present in the original code. Each indented block is part of the same top-level `||` term as its non-indented "header" line, and each corresponds to a different `token.value`.

File: compiler/src/dmd/parse.d (hunk 2 above)
Comment (pbackus) [reply compiler/src/dmd/parse.d:5015]: `function auto` is currently an error. There's an argument to be made for allowing it, but IMO it is out of scope for this PR.

**Pull request merged**